### PR TITLE
Fix/embedding sqlite

### PR DESCRIPTION
**Pull request by giauphan**
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactor (no functional change)
- [ ] Test improvement

## Checklist

- [ ] `pnpm run build` passes
- [ ] `pnpm test` passes
- [ ] Dashboard builds (`cd dashboard && pnpm run build`) if dashboard touched
- [ ] Tests added for new behavior
- [ ] Documentation updated (`README.md`, `docs/`, `CHANGELOG.md`) if user-visible change
- [ ] No secrets, credentials, or personal data in diff
- [ ] No `any` types — used `unknown` + type guards
- [ ] Commit messages follow Conventional Commits (`feat:`, `fix:`, `docs:`, etc.)

### CODE REVIEW

**Review by giauphan — COMMENTED** [review]
cc @jules

The PR looks solid and demonstrates thoughtful updates to address the `Vector Dimension Mismatch Crash` and to improve maintainability, error handling, and compatibility with SQLite and PostgreSQL. Below is the review for correctness, security, and other aspects, presented concisely:

### Code Review

#### **Correctness**
- **File: `scripts/fix-embed-dim.ts`**
  - **Line 35**: Properly checks for the existence of tables to avoid SQL errors — ✅ Safe.
  - **Lines 52-62**: Embedding truncation is calculated correctly by directly slicing the correct portion of the `ArrayBuffer` — ✅ Accurate logic. 
  - **Line 76-83**: Verification ensures the embedding was updated correctly, reducing risk of incomplete fixes — ✅ Robust validation.

- **File: `scripts/reembed-migrate.ts`**
  - **Lines 48-62**: Skips rows with null/empty content appropriately, ensuring safe processing — ✅ Defensive programming.
  - **Line 69**: Asynchronous `generateEmbedding` is handled properly, with validation for null/incorrect dimensions — ✅ Reliable.
  - **Line 88-96**: Dry run and verification logic confirm updates took effect before proceeding, promoting integrity — ✅ Well implemented.

- **File: `src/database/adapters/sqliteAdapter.ts`**
  - **Line 140**: Includes additional filtering via `length(embedding) = :expectedBytes` to handle mismatched dimension rows, preventing crashes — ✅ Correct fix.

- **File: `src/services/consolidationEngine.ts`**
  - **Line 444**: Replaced PostgreSQL-specific functions (`GREATEST`, `LEAST`) with SQLite-compatible logic (`MAX`, `MIN`) based on `dbType`. Logic seems proper for both databases — ✅ Correct fallback.

#### **Security**
- **Secrets validation**: Makes use of environment variables for database path and dimensions, avoiding hardcoded secrets — ✅ Security best practice.
- **SQL Injection**: Proper use of parameterized queries in all scripts such as `badRows` queries or the `updateStmt` — ✅ Safe.

#### **Performance**
- **Improved filtering**: Adding `length(embedding) = :expectedBytes` on the `SQLiteAdapter.searchVector()` reduces unnecessary processing of large datasets, improving query efficiency.

#### **Maintainability**
- Well-commented logic in both migration scripts and throughout the SQLite adapter — ✅ Easy to follow.
- Comprehensive use of dry-run and verification steps in `reembed-migrate.ts`, enabling safer updates and reversibility — ✅ Maintainable.

#### **Code Quality**
- Testability: Both migration scripts (`fix-embed-dim.ts`, `reembed-migrate.ts`) have clear separation of responsibilities and robust logging for debugging — ✅ Good practice.
- Redundancy: No dead code observed.
- Edge cases: Proper checks implemented (e.g., empty `content` handling in `reembed-migrate.ts`) — ✅ Thoughtful.

#### **Potential Suggestions**
1. **Error Handling**:
   - `generateEmbedding` errors (in `reembed-migrate.ts`) are logged, but you might want to classify transient vs critical failures for better reporting or retry mechanisms.

2. **Test Coverage**:
   - While not directly visible in this PR, ensure these changes are adequately tested in integration tests for specific database behaviors (e.g., truncated embeddings, invalid rows).

3. **Performance Optimization**:
   - Consider batching `UPDATE` queries in `fix-embed-dim.ts` and `reembed-migrate.ts` to reduce the number of database calls. This could be especially helpful for large-scale migrations.

4. **Logging Enhancement**:
   - Add more granular logging metrics for time taken per table/process during migrations to assess script performance.

---

### Summary
✅ Looks good overall! Key focus areas (correctness, security, and maintainability) are well-covered. The updates seem to address root causes (dimensional mismatch issues) comprehensively, with strong safeguards (dry-runs, validation). Just minor room for optional improvements around batched updates and error classification.

No critical blockers. Ready to merge.

File: scripts/fix-embed-dim.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/**
```
Comment (giauphan) [inline scripts/fix-embed-dim.ts:1]: `scripts/fix-embed-dim.ts`**

File: scripts/reembed-migrate.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/**
```
Comment (giauphan) [inline scripts/reembed-migrate.ts:1]: `scripts/reembed-migrate.ts`**

**Review by giauphan — COMMENTED** [review]
cc @jules

### Review of GitHub PR Changes:

---

#### General Observations:
The pull request applies fixes to address embedding dimension mismatches in SQLite tables and enhances the migration and consolidation logic. The changes are well-documented and follow a logical structure. Below, I analyze the code changes for correctness, security, performance, code quality, and maintainability, along with specific observations.

---

### Detailed Review:

#### Changelog (`CHANGELOG.md`):
- **Changes Overview**: Documented fixes related to dimension mismatch crashes. The update is clear, providing context on the issue addressed. No issues noted here.
  
---

#### Package Version (`package.json`):
- **Version Bump**: Updated version from `2.14.6` to `2.14.7` to mark this bug fix. No issues.

---

#### Script: `fix-embed-dim.ts`:
**Correctness**:
- The script verifies if tables exist before proceeding, ensuring database operations do not fail due to undefined references.
- Proper validation is in place to check the `blob.byteLength` and truncate if mismatched dimensions are detected.
- Truncation uses the lower-level `ArrayBuffer` approach for precision, ensuring embedding correctness.

**Security**:
- No sensitive information (e.g., credentials) is hardcoded. The database path and dimension parameters are safely retrieved through environment variables.

**Performance**:
- Efficient filtering with SQL (`length(embedding) = 16384`) minimizes unnecessary data processing.
- `prepare` statements reuse queries, improving performance.

**Code Quality**:
- There’s appropriate error handling for failed truncation or updates.
- Logs (`console.log`) provide insights during runtime, assisting debugging.

**Maintainability**:
- The use of constants like `targetDim` and `expectedBytes` makes the code adaptable for future dimension changes.

**Suggestions**:
- File: `scripts/fix-embed-dim.ts -> Line 43`: Add more context to the log message when skipping rows due to mismatched blob byte lengths.
- File: `scripts/fix-embed-dim.ts -> Line 72`: Consider wrapping `db.prepare` calls in try-catch blocks to handle unexpected SQL errors gracefully.

---

#### Script: `reembed-migrate.ts`:
**Correctness**:
- The script validates mismatched embedding lengths, ensuring consistency with the target dimension.
- The fallback to nullify embeddings for empty content prevents invalid rows from persisting in the database.

**Security**:
- The script depends on an external `generateEmbedding` service. Ensure this service has adequate security, rate limiting, and input sanitation to mitigate injection risks.
- Verify that embedding content does not expose sensitive data.

**Performance**:
- The dry-run flag (`--dry-run`) allows lightweight testing by skipping the update phase. This makes development and debugging faster.
- Utilizing prepared statements for frequent updates ensures optimized execution.

**Code Quality**:
- Detailed logs confirm individual row updates, highlighting errors during the migration process.
- Functions like `updateStmt` are reused for different rows, adhering to DRY principles.

**Maintainability**:
- Modular structure with constants like `TABLES`, `targetDim`, and SQL fragments eases edits for additional tables or dimension updates.

**Suggestions**:
- File: `scripts/reembed-migrate.ts -> Line 42`: Consider adding pagination for very large tables to avoid fetching all mismatched rows at once, which could strain memory resources.
- File: `scripts/reembed-migrate.ts -> Line 63`: Handle scenarios where `generateEmbedding` may fail multiple times (e.g., retries or fallback).
  
---

#### SQLite Adapter (`src/database/adapters/sqliteAdapter.ts`):
**Correctness**:
- Added a check for `length(embedding)` in `searchVector()` queries to avoid dimension mismatches during cosine similarity calculations. Excellent safeguard against runtime crashes.

**Security**:
- No immediately visible vulnerability in the query construction. The bindings are properly parameterized (`:queryVector`, `:expectedBytes`, etc.) to prevent SQL injection.

**Performance**:
- The filter `AND length(embedding) = :expectedBytes` ensures queries are limited to valid blobs before executing expensive cosine similarity calculations.
- Well-suited for high-scale operations when embeddings have consistent dimensions.

**Code Quality**:
- Code changes maintain readability and integrate seamlessly with the existing query structure.

**Suggestions**:
- None.

---

#### Consolidation Engine (`src/services/consolidationEngine.ts`):
**Correctness**:
- Replaced hardcoded PostgreSQL functions `GREATEST` and `LEAST` with `MAX` and `MIN` for SQLite compatibility. The changes correctly handle database-specific SQL syntax variations.

**Performance**:
- Performance for this update should not be adversely impacted, as SQLite's scalar functions are efficient.

**Suggestions**:
- File: `src/services/consolidationEngine.ts -> Lines 444-473`: While the dynamic handling of `MAX`/`MIN` is functional, future refinements might explore abstraction for cross-database SQL generation.

---

### Summary:

The pull request efficiently fixes dimension mismatch issues and improves robustness for multi-dimensional embeddings. Code enhancements show attention to database integrity, correctness, and maintainability. There are no critical issues, though areas for further enhancement around error handling and runtime efficiency were identified and suggested.

---

### Approval Status:
✅ Approve with Minor Suggestions

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 # Changelog
```
Comment (giauphan) [inline CHANGELOG.md:1]: #### Changelog (`CHANGELOG.md`):

File: scripts/fix-embed-dim.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+  console.log(`[Fix] ${table}: found ${badRows.length} 4096-dim rows to truncate to 1024-dim.`);
+
+  const updateStmt = db.prepare(`UPDATE ${table} SET embedding = ? WHERE id = ?`);
+  const verifyStmt = db.prepare(`SELECT length(embedding) as new_bytes FROM ${table} WHERE id = ?`);
+
+  for (const { id, embedding: blob } of badRows) {
+    if (!blob || blob.byteLength !== 16384) {
+      totalSkipped++;
+      continue;
```
Comment (giauphan) [inline scripts/fix-embed-dim.ts:46]: `scripts/fix-embed-dim.ts -> Line 43`: Add more context to the log message when skipping rows due to mismatched blob byte lengths.

File: scripts/fix-embed-dim.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      console.log(`[Fix] ${table}/${id}: 16384 -> ${verified.new_bytes} bytes`);
+      totalFixed++;
+    } else {
+      console.error(`[Fix] ${table}/${id}: verification failed, expected ${expectedBytes}, got ${verified.new_bytes}`);
+    }
+  }
+}
+
+console.log(`\n[Fix] Done. Fixed: ${totalFixed}, Skipped: ${totalSkipped}`);
+db.close();
```
Comment (giauphan) [inline scripts/fix-embed-dim.ts:82]: `scripts/fix-embed-dim.ts -> Line 72`: Consider wrapping `db.prepare` calls in try-catch blocks to handle unexpected SQL errors gracefully.

File: scripts/reembed-migrate.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+let totalFixed = 0;
+let totalSkipped = 0;
+let totalFailed = 0;
+
+for (const { table, contentCol, idCol } of TABLES) {
+  const exists = db.prepare("SELECT name FROM sqlite_master WHERE type='table' AND name=?").get(table);
+  if (!exists) {
+    console.log(`[Reembed] ${table}: does not exist, skipping.`);
+    continue;
+  }
```
Comment (giauphan) [inline scripts/reembed-migrate.ts:42]: `scripts/reembed-migrate.ts -> Line 42`: Consider adding pagination for very large tables to avoid fetching all mismatched rows at once, which could strain memory resources.

File: scripts/reembed-migrate.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+  const updateStmt = db.prepare(`UPDATE ${table} SET embedding = ? WHERE ${idCol} = ?`);
+  const verifyStmt = db.prepare(`SELECT length(embedding) as new_bytes FROM ${table} WHERE ${idCol} = ?`);
+
+  for (const row of badRows) {
+    if (!row.content?.trim()) {
+      console.log(`[Reembed] ${table}/${row.id}: empty content, nullifying.`);
+      if (!dryRun) updateStmt.run(null, row.id);
+      totalSkipped++;
+      continue;
```
Comment (giauphan) [inline scripts/reembed-migrate.ts:64]: `scripts/reembed-migrate.ts -> Line 63`: Handle scenarios where `generateEmbedding` may fail multiple times (e.g., retries or fallback).

**Review by giauphan — COMMENTED** [review]
cc @jules

## Review of the Pull Request

### Overview
This pull request introduces improvements and fixes primarily targeting vector embedding handling and consolidation logic within a database. Below is a categorized evaluation of the proposed changes.

---

### 1. **Correctness**
- **Fixed `vec_distance_cosine` Crash (4096 vs 1024):**
  - `sqliteAdapter.ts`: The addition of `length(embedding) = :expectedBytes` filter for queries ensures vectors of incorrect dimensions are excluded, addressing the crash. ✅
  - The correctness of the logic has been accounted for, including the calculation of the blob size \( `embedding.length * 4` \).
  
- **Scripts for Data Fixing and Migration:**
  - `fix-embed-dim.ts` and `reembed-migrate.ts`: These scripts provide effective handling of bad rows through validation, slicing operations, and retries for API calls.
  - Mitigation of edge cases:
    - Empty/null embeddings: Explicitly skipped. ✅
    - Errors in floating-point truncation or updates: Proper error handling (e.g., `try-catch` blocks and retry loops with exponential backoff). ✅
  - `fix-embed-dim.ts`: Correctly slices oversized embeddings, ensuring only the first 1024 values are retained — this is efficient.

- **Confidence Update Logic:**
  - Consolidation engine (`consolidationEngine.ts`): Adjustments for ensuring compatibility with SQLite (`MAX`/`MIN` instead of PostgreSQL's `GREATEST`/`LEAST`) make the logic more robust.

---

### 2. **Security**
- No hardcoded secrets or unsafe patterns were identified. 👍
- Example: Secrets (e.g., `COMMAND_CODE_API_KEY`) are securely passed through environment variables. ✅

---

### 3. **Performance**
- **Query Performance:**
  - Adding `length(embedding) = :expectedBytes` in queries avoids unnecessary computation on invalid rows. 👍
  - Combining filtering logic directly into the SQL avoids excessive data loading in scripts.
  
- **Batch Processing for Large Data:**
  - `reembed-migrate.ts`: Pagination via `LIMIT 1000` prevents memory bloat during processing.
  - `fix-embed-dim.ts` selectively processes only the malformed rows.

- **Retry Logic:**
  - `reembed-migrate.ts`: Exponential backoff minimizes network strain during failures.

---

### 4. **Code Quality**
- Code is modular and well-structured:
  - Scripts like `fix-embed-dim.ts` and `reembed-migrate.ts` iterate through a clearly defined pipeline of validation, transformation, and updates.
  
- Use of constants and environment variables improves maintainability:
  - Example: `targetDim`, `expectedBytes`, and `dbPath`.

- **Logging:**
  - Informative logs (e.g., before/after states, success/failure counts) enhance traceability of operations.

---

### 5. **Maintainability**
- Scripts are easy to extend:
  - E.g., Adding new tables in `TABLES` in `reembed-migrate.ts` would not require significant refactoring.
  
- **Error Handling:**
  - Structured error handling is in place to minimize runtime failures. 
  - Example: Explicit skips are logged when working with invalid or null embeddings.

---

### 6. **Specific Feedback on Code Changes**
#### `fix-embed-dim.ts`
- **Positive:**
  - The script efficiently resolves higher-dimensional embeddings without the need for API calls.
  - Validation logic prevents unnecessary processing of already-correct records.
- **Recommendation:**
  - Add unit tests covering `truncated` buffer conversion for edge cases.
  
#### `reembed-migrate.ts`
- **Positive:**
  - Comprehensive retry mechanism ensures resilience when embedding APIs fail.
- **Improvement:**
  - Consider abstracting retries as a utility function to reuse.

#### `sqliteAdapter.ts`
- **Positive:**
  - Filtering embeddings with the correct length ensures query safety and integrity.
- **Improvement:**
  - No further suggestions; the filter is concise and effective.

#### `consolidationEngine.ts`
- **Positive:**
  - Adjustments for SQLite compatibility (e.g., `MAX`/`MIN`) are useful for cross-database consistency.
- **Improvement:**
  - No significant concerns here; the changes are minimal but impactful.

---

### 7. **Changelog & Dependency Management**
- **`CHANGELOG.md`** appropriately documents:
  - Fixes and enhancements.
  - Introduction of new scripts.
- **`package.json`:**
  - Version bump to `2.14.7` is correctly included. 👍

---

### Summary of Issues Flagged
No critical issues were identified; however:
1. **Testing Recommendation:** Include unit tests for `fix-embed-dim.ts`, specifically for truncation logic.
2. **Utility Abstraction:** Retry logic in `reembed-migrate.ts` could be refactored to a shared utility to minimize duplication.

---

### Final Review

**Overall Assessment:** ✅
- This pull request addresses all points in the problem statement diligently:
  - Key fixes (e.g., dimension mismatch) and enhancements (robust logging, data migration scripts) significantly improve both correctness and maintainability.
  - No evident logic bugs or security risks were found.

**Approval Recommendation:** The PR is ready for merge.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,11 @@
...
```
Comment (giauphan) [inline CHANGELOG.md:1]: - **`CHANGELOG.md`** appropriately documents:
